<a href="https://colab.research.google.com/github/Jonathan-Oviedo/ML-23-TEA/blob/Labs/ImageClassificationProject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Image Classification Project
By: Jonathan Oviedo and Gerald Atilus

The work done for this project was done in collaboration with each member contributing an equal amount of work for each section, so the work was just added to one collab.



# Observe/Read the data

Importing our two used libraries keras and train test split

In [ ]:
from sklearn.model_selection import train_test_split
from tensorflow import keras

loading CIFAR dataset to their variables

In [ ]:
# Load the CIFAR-10 dataset and assign to variables for testing and training
(trainImages, trainLabels), (testImages, testLabels) = keras.datasets.cifar10.load_data()

Observe each variable by printing them out.

In [ ]:
#printing out train labels and images
print("trainImages:", trainImages.shape)
print("trainLabels:", trainLabels.shape)
#printing out test labels and images
print("testImages:", testImages.shape)
print("testLabels:", testLabels.shape)

trainImages: (50000, 32, 32, 3)
trainLabels: (50000, 1)
testImages: (10000, 32, 32, 3)
testLabels: (10000, 1)


Training data has 50,000 images of the shape 32,32,3

Testing data has 10,000 images of shape 32,32,3

labels only have a single column which is the label of the image

# Splitting and modifying the Data

Splitting and training the data

In [ ]:
# Split the training data into training and validation sets.
#using 80% in training and 20% in val sets
trainImages, valImages, trainLabels, valLabels = train_test_split(
    trainImages, trainLabels, test_size=0.2, random_state=42)

Shape of trainImage


In [ ]:
#used to look for shape of train image
trainImages.shape

(40000, 32, 32, 3)

added valimages to seperate line since it caused issues when looking into train image's shape

In [ ]:
#used to look for shape of val sets
valImages.shape

(10000, 32, 32, 3)

Like the training  and testing shapes previously, train and val images returned 50,000 and 10,000 images respectively

# Normalize the pixel value

we use trainImages/255 and testImages/255 so that the pixel values are normalized to 0-1 instead of 0-255

In [ ]:
# Normalize pixel values to be between 0 and 1 for training and testing images
trainImages, testImages = trainImages / 255., testImages / 255.

In [ ]:
# Normalize validation images
val_images = valImages / 255.

# Creating the CNN using keras.

creating keras model

In [ ]:
# Creating keras model
model = keras.Sequential([
#layer with 32 filters, each of size 3x3. 
#we went with relu instead of sigmoid after some testing and found it to give a better result though this 
#could be an issue on our own end.
  keras.layers.Conv2D(32, kernel_size=(3,3) ,  activation="relu", input_shape=(32, 32, 3)),
    
# max pool layer created of size 2x2 to reduce the dimension of the previous layers
 keras.layers.MaxPooling2D((2, 2)),
    
#same as above instead using 64 filters and once again using relu instead of sigmoid
 keras.layers.Conv2D(64, kernel_size=(3,3), activation="relu"),
    
#same as the previous pool layer created
 keras.layers.MaxPooling2D((2, 2)),
    
# takes the output from the layers above and puts it into 1D.
 keras.layers.Flatten(),
    
# creating connected layer with 64 units, using relu
 keras.layers.Dense(64, activation="relu"),
    
# create another layer with 10 units for the classes previously established
 keras.layers.Dense(10)
])


# summarize the created model to display the parameters of each layer created
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 30, 30, 32)        896       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 15, 15, 32)       0         
 2D)                                                             
                                                                 
 conv2d_3 (Conv2D)           (None, 13, 13, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 6, 6, 64)         0         
 2D)                                                             
                                                                 
 flatten_1 (Flatten)         (None, 2304)              0         
                                                                 
 dense_2 (Dense)             (None, 64)               

# compiling the model we created

for compiling the model we attempted two different types of SparceCategoricalCrossentropy we chose this method due to finding that multi class classification tasks is best suited for SCC.

For the optimizer we attempted two SGD and RMS we found that RMS gave a better accuracy by about 6%

In [ ]:
# Compile the model with the SGD optimizer
#sgd_optimizer = keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
#model.compile(optimizer=sgd_optimizer,
 #             loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
  #            metrics=['accuracy'])


# Compile the model with the RMSprop optimizer
rmsprop_optimizer = keras.optimizers.RMSprop(learning_rate=0.001)
model.compile(optimizer=rmsprop_optimizer,
              loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])  

# Fitting the created Model

Here we fit the model. We tried a different number of epochs and found that 10 epochs gave our best result.
5 epochs returned 65%, 15 epochs returned 62% and 10 epochs returned around 67%

In [ ]:
# train the model using 10 epochs
hist = model.fit(trainImages, trainLabels, validation_data=(valImages, valLabels), epochs=10, batch_size=256)

Epoch 1/10
157/157 [==============================] - 1s 10ms/step - loss: 0.8815 - accuracy: 0.6957 - val_loss: 176.6707 - val_accuracy: 0.5451
Epoch 2/10
157/157 [==============================] - 2s 10ms/step - loss: 0.8399 - accuracy: 0.7097 - val_loss: 179.9561 - val_accuracy: 0.5331
Epoch 3/10
157/157 [==============================] - 1s 9ms/step - loss: 0.8029 - accuracy: 0.7223 - val_loss: 246.7264 - val_accuracy: 0.4861
Epoch 4/10
157/157 [==============================] - 1s 8ms/step - loss: 0.7722 - accuracy: 0.7340 - val_loss: 186.6096 - val_accuracy: 0.5530
Epoch 5/10
157/157 [==============================] - 1s 8ms/step - loss: 0.7347 - accuracy: 0.7481 - val_loss: 194.8896 - val_accuracy: 0.5477
Epoch 6/10
157/157 [==============================] - 1s 8ms/step - loss: 0.7049 - accuracy: 0.7577 - val_loss: 234.4758 - val_accuracy: 0.5184
Epoch 7/10
157/157 [==============================] - 1s 8ms/step - loss: 0.6740 - accuracy: 0.7679 - val_loss: 214.6334 - val_accurac

# Display the model accuracy

In [ ]:
#setting/getting the model's accuracy
Accuracy = model.evaluate(testImages, testLabels)

# Print the accuracy score
print("Accuracy: ", Accuracy[1])

313/313 [==============================] - 1s 3ms/step - loss: 1.0111 - accuracy: 0.6720
Accuracy:  0.671999990940094


We were able to achieve a 67.2% accuracy score on our test data, which was the best score we were able to get when tweaking the model compile. Other attempts usually returned 55-60 so we found that our current implementation was the best we could attempt with our current setup.

We are content with our results and believe that a better methodology could probably yield a better result, but also found that a change in SCC type could also improve the accuracy score